在Upsample和interpolate函数中加入recompute_scale_factor参数

Qin-sx opened the pull request:
<!-- TemplateReference: https://github.com/PaddlePaddle/Paddle/wiki/PULL-REQUEST-TEMPLATE--REFERENCE -->
<!-- Demo: https://github.com/PaddlePaddle/Paddle/pull/24810 -->

### PR Category
<!-- One of [ User Experience | Execute Infrastructure | Operator Mechanism | CINN | Custom Device | Performance Optimization | Distributed Strategy | Parameter Server | Communication Library | Auto Parallel | Inference | Environment Adaptation ] -->

 User Experience

### PR Types
<!-- One of [ New features | Bug fixes | Improvements | Performance | BC Breaking | Deprecations | Docs | Devs | Not User Facing | Security | Others ] -->

Improvements 

### Description
<!-- Describe what you’ve done -->

Upsample参数实际调用了interpolate参数。  
参考Pytorch，recompute_scale_factor为True时计算output size，然后重置scale参数。  

API文档修改：https://github.com/PaddlePaddle/docs/pull/7205
PaConvert修改：https://github.com/PaddlePaddle/PaConvert/pull/567

Qin-sx committed:
added recompute_scale_factor in interpolate

	modified:   python/paddle/nn/functional/common.py
	modified:   python/paddle/nn/layer/common.py
	new file:   test/legacy_test/test_interp_recompute_scale_factor.py

Qin-sx committed:
added an error test

	modified:   test/legacy_test/test_interp_recompute_scale_factor.py

Qin-sx committed:
added comments for new para

	modified:   python/paddle/nn/functional/common.py
	modified:   python/paddle/nn/layer/common.py

Qin-sx committed:
Merge branch 'develop' into opt-Upsample

Qin-sx commented:
[对应文档修改](https://github.com/PaddlePaddle/docs/pull/7205)

Qin-sx committed:
Extend timeout of test

	modified:   test/legacy_test/CMakeLists.txt

zhwesky2010 commented on python/paddle/nn/functional/common.py:
```diff
@@ -493,6 +500,15 @@ def interpolate(
 
+    if (
+        recompute_scale_factor is not None
+        and recompute_scale_factor
```
这几个条件语句可以优化下写法，使其更易读一些

Qin-sx replied:
收到，已修改。

zhwesky2010 reviewed (commented):
paconvert也加下测试，在paconvert/tests目录，这个会和pytorch的运行结果作对比

Qin-sx committed:
added more tests for coverage

	modified:   test/legacy_test/test_interp_recompute_scale_factor.py

Qin-sx committed:
simplified conditions

	modified:   python/paddle/nn/functional/common.py

Qin-sx commented:
[PaConvert对应PR](https://github.com/PaddlePaddle/PaConvert/pull/567)

zhwesky2010 commented on python/paddle/nn/functional/common.py:
```diff
@@ -493,6 +500,11 @@ def interpolate(
             "align_corners option can only be set with the interpolating modes: linear | bilinear | bicubic | trilinear"
         )
 
+    if recompute_scale_factor and size is not None:
```
这个放到下面的
```
if out_shape is not None:
```
分支下吧

Qin-sx replied:
收到，已修改。

zhwesky2010 commented on python/paddle/nn/functional/common.py:
```diff
@@ -645,7 +657,7 @@ def _is_list_or_tuple_(data):
                     attrs['out_w'] = out_shape[2]
 
-    else:
+    elif scale is not None and recompute_scale_factor is not True:
```
分支逻辑梳理一下，是否是：
```
if out_shape is not None:
    pass
elif scale is not None:
    if recompute_scale_factor:
        pass
    else:
        pass
else:
    raise error
```

Qin-sx replied:
收到，已按此修改。

Qin-sx committed:
optimized conditions

	modified:   python/paddle/nn/functional/common.py

zhwesky2010 commented on python/paddle/nn/functional/common.py:
```diff
@@ -550,8 +562,6 @@ def _is_list_or_tuple_(data):
 
     out_shape = size
     scale = scale_factor
-    if out_shape is not None and scale is not None:
```
这个为什么删除？

zhwesky2010 commented on python/paddle/nn/functional/common.py:
```diff
@@ -493,6 +500,11 @@ def interpolate(
             "align_corners option can only be set with the interpolating modes: linear | bilinear | bicubic | trilinear"
         )
 
+    if recompute_scale_factor and size is not None:
```
这个并没有修改

Qin-sx replied:
抱歉，这个上面的看错了，已修改。

Qin-sx committed:
updated conditions again

	modified:   python/paddle/nn/functional/common.py

zhwesky2010 commented on test/legacy_test/CMakeLists.txt:
```diff
@@ -833,6 +833,7 @@ set_tests_properties(test_multiprocess_dataloader_iterable_dataset_static
 set_tests_properties(test_lstm_cudnn_op PROPERTIES TIMEOUT 120)
 set_tests_properties(test_stack_op PROPERTIES TIMEOUT 120)
 set_tests_properties(test_bilinear_interp_v2_op PROPERTIES TIMEOUT 120)
+set_tests_properties(test_interp_recompute_scale_factor PROPERTIES TIMEOUT 120)
```
这个耗时比较长，有办法精简下case缩短下时间不？

Qin-sx committed:
reduced tests

	modified:   test/legacy_test/CMakeLists.txt
	modified:   test/legacy_test/test_interp_recompute_scale_factor.py

Qin-sx committed:
Merge branch 'develop' into opt-Upsample

zhwesky2010 reviewed (approved):
LGTM

sunzhongkai588 reviewed (approved):
LGTM
@Qin-sx 中文文档也同步修改

Qin-sx commented:
> LGTM @Qin-sx 中文文档也同步修改

您好，中文文档应该已经修改了  
https://github.com/PaddlePaddle/docs/pull/7205

luotao1 commented:
请merge下develop，重新跑下CI

Qin-sx committed:
Merge branch 'develop' into opt-Upsample
